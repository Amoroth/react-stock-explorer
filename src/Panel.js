import React from 'react'
import { css } from 'emotion'

const panel = (props) => {
  let priceChange = `${props.change > 0 ? '+' : ''}${(props.change * 100).toFixed(2)}%`
  let stockName = props.title
  if (stockName.length > 30)
  stockName = stockName.slice(0, 30) + '...'
  return (
    <div className={style}>
      <h6 className={titleStyle} title={ stockName[stockName.length - 2] === '.' ? props.title : null }>{ stockName }
      </h6>
      <div>
        <span>{ props.short }</span>
        <span>{ props.price }</span>
        {/* <span>Something?</span> */}
        <span 
          style={{color: props.change > 0 ? 'green' : 'red'}}>
          { priceChange }
        </span>
      </div>
    </div>
  )
}

const style = css({
  backgroundColor: '#fff',
  borderRadius: '5px',
  boxShadow: '2px 4px 10px rgba(0,0,0,.0125)',
  fontSize: '15px',
  margin: '4px 4px 4px 4px',
  maxWidth: '720px',
  padding: '16px',
  transition: 'box-shadow .1s ease-in-out',
  width: '48%',
  'div': {
    display: 'flex',
    justifyContent: 'space-between'
  },
  '&:hover': {
    boxShadow: '2px 4px 10px rgba(0,0,0,.1)',
  },
  '@media screen and (max-width: 800px)': {
    width: '60%',
    maxWidth: '97%'
  },
  '@media screen and (max-width: 620px)': {
    width: '100%',
    maxWidth: '97%'
  },
})

const titleStyle = css({
  fontSize: '21px',
  fontWeight: 'bold',
  margin: '0 0 10px 0',
})

const shortStyle = css({
  fontSize: '15px',
  fontWeight: 'normal',
  marginLeft: '8px',
})

export default panel