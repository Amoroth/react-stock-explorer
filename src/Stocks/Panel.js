import React from 'react'
import PropTypes from 'prop-types'
import { withRouter } from 'react-router-dom'

import styles from './Stocks.module.css'

const panel = ({ change, title, short, favorite, onFav, price, history }) => {
  const priceChange = `${change > 0 ? '+' : ''}${(change * 100).toFixed(2)}%`
  let stockName = title
  if (stockName.length > 30) {
    stockName = `${stockName.slice(0, 28)}...`
  }

  const openFullPage = (e) => {
    if ((e.type === 'keypress' && e.which === 13) || e.type === 'click') {
      history.push(`/stock?cmp=${short}`)
    }
  }

  return (
    <div
      className={styles.panel}
      onClick={openFullPage}
      role="button"
      onKeyPress={openFullPage}
      tabIndex={0}
    >
      <div className={styles['panel-upper']}>
        <h6
          className={styles['panel-title']}
          title={stockName[stockName.length - 2] === '.' ? title : null}
        >
          {stockName}
        </h6>
        <button
          className={styles['panel-favorite']}
          onClick={(e) => onFav(e, short)}
          type="button"
        >
          {favorite ? (
            <i className="material-icons" style={{ color: 'red' }}>
              favorite
            </i>
          ) : (
            <i className="material-icons">favorite_border</i>
          )}
        </button>
      </div>
      <div className={styles['panel-details']}>
        <span>{short}</span>
        <span>{`${price} USD`}</span>
        <span style={{ color: change > 0 ? 'green' : 'red' }}>
          {priceChange}
        </span>
      </div>
    </div>
  )
}

panel.propTypes = {
  change: PropTypes.number.isRequired,
  title: PropTypes.string.isRequired,
  short: PropTypes.string.isRequired,
  favorite: PropTypes.bool,
  onFav: PropTypes.func.isRequired,
  history: PropTypes.shape({ push: PropTypes.func.isRequired }).isRequired,
  price: PropTypes.number.isRequired,
}

panel.defaultProps = { favorite: false }

export default withRouter(panel)
