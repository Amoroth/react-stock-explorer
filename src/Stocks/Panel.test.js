import React from 'react'
import { shallow } from 'enzyme'

import Panel from './Panel'

describe('<Panel />', () => {
  let props
  let wrapper

  beforeEach(() => {
    props = {
      change: 0.04,
      title: 'Company',
      short: 'COM',
      favorite: false,
      onFav: jest.fn(),
      history: { push: jest.fn() },
      price: '4',
      currency: 'USD',
    }
    wrapper = shallow(<Panel.WrappedComponent {...props} />)
  })

  it('renders', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('formats and styles percentage', () => {
    let percentage = wrapper.find('.panel-details').children().last()

    expect(percentage.text()).toBe('+4.00%')
    expect(percentage.prop('style')).toEqual({ color: 'green' })

    wrapper.setProps({ change: -0.02 })
    percentage = wrapper.find('.panel-details').children().last()

    expect(percentage.text()).toBe('-2.00%')
    expect(percentage.prop('style')).toEqual({ color: 'red' })
  })

  it('opens full page', () => {
    wrapper.simulate('click', { type: 'click' })

    expect(props.history.push).toBeCalledWith(`/stock?cmp=${props.short}`)
  })

  it('opens full page with keyboard', () => {
    wrapper.simulate('keypress', { type: 'keypress', which: 13 })

    expect(props.history.push).toBeCalledWith(`/stock?cmp=${props.short}`)
  })

  it('shortens title', () => {
    expect(wrapper.find('h6').text()).toBe(props.title)

    wrapper.setProps({ title: 'Another company created by Elon Musk' })

    expect(wrapper.find('h6').text()).toBe('Another company created by E...')
  })

  it('h6 has title prop if title is too long', () => {
    expect(wrapper.find('h6').prop('title')).toBe(null)

    const longName = 'Another company created by Elon Musk'
    wrapper.setProps({ title: longName })

    expect(wrapper.find('h6').prop('title')).toBe(longName)
  })

  it('saves company on favorite', () => {
    wrapper.find('.panel-favorite').simulate('click')

    expect(props.onFav).toBeCalled()
    expect(props.onFav.mock.calls[0][1]).toBe(props.short)
  })

  it('styles favorite button', () => {
    expect(wrapper.find('.panel-favorite i').text()).toBe('favorite_border')

    wrapper.setProps({ favorite: true })

    expect(wrapper.find('.panel-favorite i').text()).toBe('favorite')
    expect(wrapper.find('.panel-favorite i').prop('style')).toEqual({ color: 'red' })
  })

  it('sets price to 0 if NaN', () => {
    expect(wrapper.find('.panel-details').childAt(1).text()).toBe('4 USD')

    wrapper.setProps({ price: 'NaN' })

    expect(wrapper.find('.panel-details').childAt(1).text()).toBe('0 USD')
  })
})
