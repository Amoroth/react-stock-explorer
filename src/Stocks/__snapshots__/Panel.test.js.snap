// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Panel /> renders 1`] = `
<div
  className="panel"
  onClick={[Function]}
  onKeyPress={[Function]}
  role="button"
  tabIndex={0}
>
  <div
    className="panel-upper"
  >
    <h6
      className="panel-title"
      title={null}
    >
      Company
    </h6>
    <button
      className="panel-favorite"
      onClick={[Function]}
      type="button"
    >
      <i
        className="material-icons"
      >
        favorite_border
      </i>
    </button>
  </div>
  <div
    className="panel-details"
  >
    <span>
      COM
    </span>
    <span>
      4 USD
    </span>
    <span
      style={
        Object {
          "color": "green",
        }
      }
    >
      +4.00%
    </span>
  </div>
</div>
`;
